### DISCUSSION TITLE
How to catch exceptions in my custom exception middleware?

### DISCUSSION THREAD

**Discussion by bwn-z** [ThreeMammals/Ocelot#1728] [Q&A]
## Expected Behavior / New Feature
I want to add my custom exceptionMiddleware and catch all exception in it. 
I want to do this because I don't like the default error response (statusCodes: 400+, 500+) from ocelot without json body and info.

[Obviously you can just add middleware as normal before the call to app.UseOcelot() It cannot be added after as Ocelot does not call the next middleware.](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html)

## Actual Behavior / Motivation for New Feature
I added my custom exceptionMiddleware and It doesn't get the exceptions after DelegatingHandler callings.

## Steps to Reproduce the Problem
1. Register at Startup.cs an exceptionMiddleware before a registration of Ocelot
`app.UseMiddleware<ExceptionMiddleware>()`;
`app.UseOcelot().Wait();`
2. Register a some delegating handler at `public void ConfigureServices(IServiceCollection services)`:
  `services.AddOcelot(Configuration).AddDelegatingHandler<SomeDelegatingHandler>();` 
3. Use at ocelot.json the configuration: `"DelegatingHandlers" : [ "SomeDelegatingHandler" ]` for a some route.
4.  ExcetionMiddleware.cs code:
```csharp
public async Task InvokeAsync(HttpContext context) 
{
    try 
    {
        await _next(context); // <-- set the breakpoint here
    } 
    catch (Exception ex) 
    {
        Console.WriteLine(ex); // <-- set the breakpoint here
    }
}
```
5. SomeDelegatingHandler.cs code:
```csharp
protected override Task < HttpResponseMessage > SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) 
{
    throw new Exception("test exception"); // <-- set the breakpoint here
    return base.SendAsync(request, cancellationToken);
}
```
6. OcelotPipeline catches, handles and writes a log of the error. We can't get them in our exceptionMiddleware. Manipulations with OcelotPipelineConfiguration and its middlewares (**_PreErrorResponderMiddleware PreAuthenticationMiddleware AuthenticationMiddleware, PreAuthorizationMiddleware, AuthorizationMiddleware, PreQueryStringBuilderMiddleware_**) with try/catch design don't work too.

## Specifications

  - Version: 17.0.1
  - Platform: net5.0
  - Subsystem: Windows 10 x64

**Comment by ks1990cn** [comment]
@bwn-z  I don't think so this is even possible, because:
You are creating error middleware and ocelot is also a middleware, when request comes in it get caught by ocelot and this is bad way to call previous or specific middleware on current middleware, this may destroy middleware chaining

Anything breaks with ocelot, everything comes with message and status codes defined

@ggnaegi @raman-m can you please check my thoughts  & close this if my answer sounds correct?

**Comment by raman-m** [comment]
@bwn-z 
Thanks for your interest in Ocelot!
How many years do you write code as ASP.NET developer? Learn more about [UnhandledException](https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.unhandledexception?view=net-7.0#system-appdomain-unhandledexception) please!

Your idea to catch all exceptions in Ocelot is wrong! Don't try that because it is impossible.
Currently Ocelot catches all unhandled exceptions by [ExceptionHandlerMiddleware](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20ExceptionHandlerMiddleware&type=code).  This middleware is attached to Ocelot pipeline forever. I'm not going to explain you why Ocelot catches all unhandled exception. There are a lot of reasons to do that.
This design will remain forever. So, we are not going no switch off this `ExceptionHandlerMiddleware`. Design is not good, it has strange logic, but it works.

Regarding having the exception object. It is impossible to catch this object now. But in theory, we can enhance `ExceptionHandlerMiddleware` to introduce events model for consuming. It makes sense...
If you subscribe to those events then you will have exception object to process in subscriber's context.
But this is a coding challenge. We have to discuss this events model, and convert to a real issue for development.

Again, imagine you got exception object. And what are you going to do with this object?
**What's your goal while catching all exceptions?**
To log message to somewhere? But it is logged by standard Ocelot logger.
To generate some event? But for what purposes? Who will be a consumer if the event will come outside of AppDomain?

**Comment by raman-m** [comment]
@bwn-z 
if you want to catch all exceptions and manage these objects somehow, you have to fork Ocelot repository to your account.
After forking just develop [ExceptionHandlerMiddleware](../blob/develop/src/Ocelot/Errors/Middleware/ExceptionHandlerMiddleware.cs) class for your needs.
Enjoy!

**Reply by ks1990cn** [reply]
If they gonna customize code for generic status code and message, will they be able to report our issue if they face?

I agree its open source and everyone have right to think broad, just a thought if they gonna really be able to debug.

**Comment by ggnaegi** [comment]
> After forking just develop [ExceptionHandlerMiddleware](../blob/develop/src/Ocelot/Errors/Middleware/ExceptionHandlerMiddleware.cs) class for your needs.

Yes, this will work for sure! 😄 🚀 

Or, maybe, you could use the extension method ```UseStatusCodePages``` and put it before ocelot middlewares and write the response accordingly.
https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.statuscodepagesextensions.usestatuscodepages?view=aspnetcore-7.0

It's just an idea, I haven't tried it.
```c#
app.UseStatusCodePages(new StatusCodePagesOptions
      {
          HandleAsync = async ctx =>
          {
              if (ctx.HttpContext.Response.StatusCode == 404)
              {
                  await ctx.HttpContext.Response.WriteAsync("Uuuups, path not found...");
              }
          }
      });
```

**Reply by raman-m** [reply]
Wow! This is cool feature!
Usage of `UseStatusCodePages()` is better than writing a custom middleware to process exceptions. Definitely better approach.

@bwn-z Do you like it?
